### ISSUE TITLE
ARM hyperkube broken at v1.4.0-alpha.2

### ISSUE THREAD

**Issue by luxas** [kubernetes/kubernetes#29904]
v1.4.0-alpha.2 linux/arm hyperkube build is broken:
```
$ /hyperkube
fatal error: newproc: function arguments too large for new goroutine

runtime stack:
runtime.throw(0x2ed1850, 0x37)
	/usr/local/go/src/runtime/panic.go:547 +0x78
runtime.newproc1(0x1487e320, 0x14783edc, 0x27a14a8, 0x0, 0x207e530, 0x14783e44)
	/usr/local/go/src/runtime/proc.go:2685 +0xa4
runtime.newproc.func1()
	/usr/local/go/src/runtime/proc.go:2661 +0x44
runtime.systemstack(0x43b5100)
	/usr/local/go/src/runtime/asm_arm.s:247 +0x80
runtime.mstart()
	/usr/local/go/src/runtime/proc.go:1051

goroutine 1 [running, locked to thread]:
runtime.systemstack_switch()
	/usr/local/go/src/runtime/asm_arm.s:192 +0x4 fp=0x14783eac sp=0x14783ea8
runtime.newproc(0x27a14a8, 0x1487e320)
	/usr/local/go/src/runtime/proc.go:2662 +0x5c fp=0x14783ed0 sp=0x14783eac
k8s.io/kubernetes/pkg/apis/abac/v1beta1.init.1()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apis/abac/v1beta1/register.go:29 +0x34 fp=0x14783f14 sp=0x14783ed0
k8s.io/kubernetes/pkg/apis/abac/v1beta1.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apis/abac/v1beta1/types.go:60 +0x54 fp=0x14783f18 sp=0x14783f14
k8s.io/kubernetes/pkg/apis/abac/latest.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apis/abac/latest/latest.go:23 +0x54 fp=0x14783f1c sp=0x14783f18
k8s.io/kubernetes/pkg/auth/authorizer/abac.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/auth/authorizer/abac/abac.go:234 +0x64 fp=0x14783f20 sp=0x14783f1c
k8s.io/kubernetes/pkg/apiserver.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/watch.go:274 +0xcc fp=0x14783f40 sp=0x14783f20
k8s.io/kubernetes/pkg/genericapiserver.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/genericapiserver/tunneler.go:204 +0x98 fp=0x14783f74 sp=0x14783f40
k8s.io/kubernetes/federation/cmd/federation-apiserver/app.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/federation/cmd/federation-apiserver/app/server.go:173 +0x54 fp=0x14783f78 sp=0x14783f74
main.init()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/cmd/hyperkube/server.go:77 +0x4c fp=0x14783f7c sp=0x14783f78
runtime.main()
	/usr/local/go/src/runtime/proc.go:177 +0x274 fp=0x14783fa4 sp=0x14783f7c
runtime.goexit()
	/usr/local/go/src/runtime/asm_arm.s:990 +0x4 fp=0x14783fa4 sp=0x14783fa4

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_arm.s:990 +0x4

goroutine 5 [chan receive]:
k8s.io/kubernetes/vendor/github.com/golang/glog.(*loggingT).flushDaemon(0x43b4d60)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:879 +0x60
created by k8s.io/kubernetes/vendor/github.com/golang/glog.init.1
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/golang/glog/glog.go:410 +0x2cc
```

Notable info:
 - linux/386 and linux/arm64 works (so it's arm-specific)
 - binaries like apiserver, federation-apiserver works on arm
 - It is probably not related to `pkg/apis/abac`, as it hasn't changed since v1.4.0-alpha.1
 - It happens on `init()` (static initialization of a lot of packages)
 - http://stackoverflow.com/questions/28225629/golang-error-function-arguments-too-large-for-new-goroutine
 - https://golang.org/src/runtime/proc.go#L2684

My guess would be that something changed in `pkg/runtime` which makes it overflow on ARM

I'm very thankful for your thoughts and help.
This should be fixed until the next alpha release. Adding Priority/P1

@smarterclayton @thockin @lavalamp @vishh @brendandburns @rcs @david-mcmahon @wojtek-t

**Comment by luxas** [comment]
I really need some Go guru to help here, maybe @smarterclayton @jfrazelle @rsc @davecheney and/or @wojtek-t 

also cc @kubernetes/sig-api-machinery

**Comment by davecheney** [comment]
What's the line of code that triggers this?

**Comment by luxas** [comment]
@davecheney https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/abac/v1beta1/register.go#L29

**Comment by davecheney** [comment]
looks like that call got inlined into its caller. What is the line of code it's calling?

**Comment by luxas** [comment]
@davecheney Sorry, I can't follow that statement.
What do you mean?

`api.Scheme.AddKnownTypes(GroupVersion, &Policy{})` is the call that's failing, but it hasn't changed since v1.4.0-alpha.1 where it was working...

https://github.com/kubernetes/kubernetes/blob/master/pkg/runtime/scheme.go#L160 is the function that's called.

**Comment by smarterclayton** [comment]
Each API group package is fairly large - protobuf, generated copies, and
generated conversions.  Is there anything about the goroutines used for
package init that passes information about the package through in a way
that may result in this error?  Is the init method itself inlined into a
parent init module by the compiler?

**Comment by davecheney** [comment]
It looks like the init method is being inlined.

**Comment by luxas** [comment]
@smarterclayton @davecheney So it might be possible to fix this issue by killing most `func init`s?
I think that would be a good idea anyway, especially if we're thinking about transistioning to hyperkube even more.

@smarterclayton Do you have any preferred place in the code to init those apis instead?

**Comment by davecheney** [comment]
On Mon, Aug 8, 2016 at 3:58 PM, Lucas Käldström <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> @davecheney
> <https://github.com/davecheney> So it might be possible to fix this issue
> by killing most func inits?
>

You should kill init functions, they're hard to test, but I have no idea if
this will fix the problem.

I recommend you write a small, selfcontained repro, it should be pretty
easy, some struct type with a large array (not a slice) passed by value
should do it. When you have this repro raise an issue on golang/go

**Comment by lavalamp** [comment]
I'm working on killing, or at least mortally wounding, those init
functions, but right now that'd be a very tall order.

**Comment by luxas** [comment]
@lavalamp Are you working on killing the inits in time for v1.4?

@smarterclayton Do you have any preference where the API registration would be instead?

**Comment by lavalamp** [comment]
It's not looking good.

**Comment by lavalamp** [comment]
Kicking this out of CSI team, not sure where it should go. If this is blocked on my cleanup, then it's definitely not making 1.5.

**Comment by pwittrock** [comment]
@mikedanese @alex-mohr Can you comment on if this is a release blocker or who owns hyperkube?  This seems like it might be a big deal, but I don't know enough about HK.

I am leaving this in the 1.4 release as a blocking issue.  Please provide daily status updates to this issue or move it out of the release and into v1.4-nonblocking where it will no longer be tracked as a release blocking issue in the burndown meetings.

**Comment by mikedanese** [comment]
It's owned by no one and everyone but we've had a bunch of problems with it recently. This shouldn't block v1.4 because no one is working on it.

**Comment by mml** [comment]
Hyperkube needs clear ownership.  Failing that, it's at least a product of the build/release toolchain.

**Comment by luxas** [comment]
Yes, we should open an hyperkube OWNERS issue after v1.4 is shipped.

@pwittrock I marked this as a release-blocker.
Together with the Golang team, I've been given a patch that fixes this issue, and I have a PR for Kubernetes in today or tomorrow that fixes this issue.

**Comment by pwittrock** [comment]
Alright sounds good.  Please update this issue tomorrow with the progress.  Please make sure the PR gets cherrypicked.

**Comment by pwittrock** [comment]
@luxas By release-blocker, you mean we would push the release date out a week if necessary in order to get this in?

**Comment by luxas** [comment]
No, would it do that if we're getting my PR in today or tomorrow?

**Comment by luxas** [comment]
@pwittrock Ok, seems like I misunderstood the full purpose of the release-blocker label, removing it.
But still, I'd like to get the PR merged

**Comment by pwittrock** [comment]
@luxas Looks like you made it in under the wire.  I will be cutting a beta tonight with your cherrypick in it.  On Monday, please manually verify the issue is resolved and then close this issue.

**Comment by pwittrock** [comment]
@luxas Can I mark this as resolved?

**Comment by luxas** [comment]
@pwittrock Yes, you can.
Now when I tested it, unfortunately another issue was uncovered (known `sync/atomic` 32-bit golang issue indeed), but I'm glad this is in, it's a step further.

Will open separate issues for that and continue working with the Golang team.
Also things will get better when we have Go 1.7 officially.

I'll probably send a patch for it in time for 1.4.1 (just moving a struct field to the top of the struct)
But thanks for cherrypicking the PR, it helped indeed.

**Comment by pwittrock** [comment]
@luxas Thanks for all your work in this area.
